build_runner_core/lib/src/asset_graph/node.dart

```diff
@@ -79,47 +84,106 @@ class AssetNode {
   ///
   /// They are "inputs" to the entire build, so they are never explicitly
   /// tracked as inputs.
-  AssetNode.internal(this.id, {this.lastKnownDigest})
-    : type = NodeType.internal;
+  AssetNode.internal(this.id, {Digest? lastKnownDigest})
+    : type = NodeType.internal,
+      _primaryOutputs = {},
+      _outputs = {},
+      _anchorOutputs = {},
+      _lastKnownDigest = lastKnownDigest,
+      _deletedBy = {};
 
   /// A manually-written source file.
-  AssetNode.source(this.id, {this.lastKnownDigest}) : type = NodeType.source;
+  AssetNode.source(
+    this.id, {
+    Digest? lastKnownDigest,
+    Iterable<AssetId>? outputs,
+    Iterable<AssetId>? primaryOutputs,
+  }) : type = NodeType.source,
+       _primaryOutputs = primaryOutputs?.toSet() ?? {},
+       _outputs = outputs?.toSet() ?? {},
+       _anchorOutputs = {},
+       _lastKnownDigest = lastKnownDigest,
+       _deletedBy = {};
 
   /// A [BuilderOptions] object.
   ///
   /// Each [GeneratedAssetNode] has one describing its configuration, so it
   /// rebuilds when the configuration changes.
-  AssetNode.builderOptions(this.id, {this.lastKnownDigest})
-    : type = NodeType.builderOptions;
+  AssetNode.builderOptions(this.id, {Digest? lastKnownDigest})
+    : type = NodeType.builderOptions,
+      _primaryOutputs = {},
+      _outputs = {},
+      _anchorOutputs = {},
+      _lastKnownDigest = lastKnownDigest,
+      _deletedBy = {};
 
   /// A missing source file.
   ///
   /// Created when a builder tries to read a non-existent file.
   ///
   /// If later the file does exist, the builder must be rerun as it can
   /// produce different output.
-  AssetNode.missingSource(this.id, {this.lastKnownDigest})
-    : type = NodeType.syntheticSource;
+  AssetNode.missingSource(this.id, {Digest? lastKnownDigest})
+    : type = NodeType.syntheticSource,
+      _primaryOutputs = {},
+      _outputs = {},
+      _anchorOutputs = {},
+      _lastKnownDigest = lastKnownDigest,
+      _deletedBy = {};
 
   /// Placeholders for useful parts of packages.
   ///
   /// Four types of placeholder are used per package: the `lib` folder, the
   /// `test` folder, the `web` folder, and the whole package.
   ///
   /// TODO(davidmorgan): describe how these are used.
-  AssetNode.placeholder(this.id, {this.lastKnownDigest})
-    : type = NodeType.placeholder;
+  AssetNode.placeholder(this.id, {Digest? lastKnownDigest})
+    : type = NodeType.placeholder,
+      _primaryOutputs = {},
+      _outputs = {},
+      _anchorOutputs = {},
+      _lastKnownDigest = lastKnownDigest,
+      _deletedBy = {};
 
   @override
   String toString() => 'AssetNode: $id';
+
+  /// Write access to collections in the node.
+  AssetNodeMutator get mutate => AssetNodeMutator(this);
+
+  /// Access to unmodifable views on collections in the node.
+  AssetNodeInspector get inspect => AssetNodeInspector(this);
+}
+
+/// Write access to collections in the node.
+extension type AssetNodeMutator(AssetNode node) {
```

Comment:
Is this supposed to do something additional in the future?
As this is now I don't think it actually makes any change, or am I missing something?

Reply:
Right, no actual change.

Rather a dev usability change, I can now search for uses of `mutate` :)

Reply:
I was also wondering what the point was - maybe put that in a comment  on the extension type